From: Diego Escalante Urrelo <diegoe@gnome.org>
Date: Thu, 13 Feb 2025 00:19:36 -0500
Subject: wl: Add explanation for IBT warnings in newer kernels

Because of the binary blob the driver will trigger the kernel's
"Unpatched return thunk in use" warning. We can not fix without
recompiling the blob.

Considering that this driver handles wifi cards that are mostly
attached/locked to pre-2020 CPUs, this patch adds a non-critical warning
message explaining the scary kernel traces and suggesting `ibt=off`.

Based on Andreas Beckmann <anbe@debian.org> patch and suggestion.

See: https://salsa.debian.org/nvidia-team/nvidia-graphics-drivers/-/blob/470/debian/module/debian/patches/0033-refuse-to-load-legacy-module-if-IBT-is-enabled.patch?ref_type=heads
See: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1084853#10
Fixes: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1084853
---
 amd64/src/wl/sys/wl_linux.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/amd64/src/wl/sys/wl_linux.c b/amd64/src/wl/sys/wl_linux.c
index 4d618b7..d56bf2c 100644
--- a/amd64/src/wl/sys/wl_linux.c
+++ b/amd64/src/wl/sys/wl_linux.c
@@ -930,6 +930,10 @@ wl_module_init(void)
 {
 	int error = -ENODEV;
 
+#ifdef CONFIG_X86_KERNEL_IBT
+	printk(KERN_WARNING "wl: This driver includes a binary blob incompatible with IBT protection, available since Intel Core Tiger Lake (11th gen, 2020). If your CPU is older you can ignore the 'Unpatched return thunk in use' warnings caused by this driver. You can disable IBT by adding `ibt=off` to your kernel boot options.");
+#endif
+
 #ifdef BCMDBG
 	if (msglevel != 0xdeadbeef)
 		wl_msg_level = msglevel;
